File: DSA/Arrays/javascript/array.js
#!/bin/node
```Array Data Structure Module```;

class Array {
  constructor(size) {
    this.size = size;
    this.length = 0;
    this.data = new Array(size).fill(null);
  }

  display() {
    ```Class method to Display the Elements```;
    if (this.length == 0) {
      console.log("Array is Empty.");
    } else {
      console.log(this.data.slice(0, this.length).join(" "));
    }
  }

  append(value) {
    ```Class method to add an Element to the end.```;
    if (this.length < this.size) {
      this.data[this.length] = value;
      this.length += 1;
    } else {
      console.log("Array is Full! Cannot add more Elements");
    }
  }

  insert(index, value) {
    ```Class method to insert an Element in a position.```;
    if (this.length >= this.size) {
      console.log("Array is full! Cannot insert.");
      return;
    }

    if (index < 0 && index > this.length) {
      console.log("Invalid index! Must be between 0 and", this.length);
      return;
    }

    for (let i = this.length; i > index; i--) {
      this.data[i] = this.data[i - 1];
    }

    this.data[index] = value;
    this.length += 1;
  }

  delete(index) {
    ```Class method to Delete an Element at Position.```;
    if (this.length === 0) {
      console.log("Array is empty! Nothing to delete.");
      return;
    }

    if (index < 0 && index >= this.length) {
      console.log("Invalid index! Must be between 0 and", this.length - 1);
      return;
    }

    for (let i = index; i < this.length - 1; i++) {
      this.data[i] = this.data[i + 1];
    }

    this.data[this.length - 1] = null;
    this.length -= 1;
  }

  search(value) {
    ```Class method to search/traverse Elements.```;
    for (let i = 0; i < this.length; i++) {
      if (this.data[i] === value) {
        return i;
      }
    }

    return -1;
  }
}

File: DSA/Arrays/python/array.py
#!/bin/python3
"""Array Data Structure Module"""

class Array:
    """Array ADT"""
    def __init__(self, size):
        """Initializing the array with a fixed size."""
        self.size = size
        self.length = 0
        self.data = [None] * size

    def display(self):
        """Class method to display the Array"""
        if self.length == 0:
            print("Array is Empty.")
        else:
            for i in range(self.length):
                print(self.data[i], end=" ")
                print()

    def append(self, value):
        """Class method to append an element to the Array."""
        if self.length < self.size:
            self.data[self.length] = value
            self.length += 1
        else:
            print("Array is Full! Cannot append more Elements.")

    def insert(self, index, value):
        """Class method to insert an Element."""
        if self.length >= self.size:
            print("Array is full! Cannot insert more Elements.")
            return

        if index < 0 or index > self.length:
            print(f"Invalid Index! must be between 0 and {self.length}")
            return

        # Shifting Elements to the Right.
        for i in range(self.length, index, -1):
            self.data[i] = self.data[i - 1]

        self.data[index] = value
        self.length += 1

    def delete(self, index):
        """Class method to delete from Array."""
        if self.length == 0:
            print("Array is empty! nothing to Delete.")
            return

        if index < 0 or index >= self.length:
            print(f"Invalid index! Must be between o and {self.length}")
            return

        # Shifting Elements to the Left.
        for i in range(index, self.length - 1):
            self.data[i] = self.data[i+1]

        # Set the last Element to None and decrease Length.
        self.data[self.length - 1] = None
        self.length -= 1

    def search(self, value):
        """Class method to search for an Element in Array."""
        for i in range (self.length):
            if self.data[i] == value:
                return i
        return -1
